src/hmem.c

```diff
@@ -126,3 +131,54 @@ ssize_t ofi_copy_to_hmem_iov(const struct iovec *hmem_iov,
 				     hmem_iov_offset, src, size,
 				     OFI_COPY_BUF_TO_IOV);
 }
+
+int ofi_get_hmem_iface(const void *buf, enum fi_hmem_iface *iface)
+{
+#ifdef HAVE_LIBCUDA
+	CUresult cuda_ret;
+	unsigned int data;
+
+	cuda_ret = cuPointerGetAttribute(&data,
+					 CU_POINTER_ATTRIBUTE_MEMORY_TYPE,
+					 (CUdeviceptr)buf);
+	switch (cuda_ret) {
+	case CUDA_SUCCESS:
+		if (data == CU_MEMORYTYPE_DEVICE) {
+			*iface = FI_HMEM_CUDA;
+			return FI_SUCCESS;
+		}
+		break;
+
+	/* Returned if the buffer is not associated with the CUcontext support
+	 * unified virtual addressing. Since host buffers may fall into this
+	 * category, this is not treated as an error.
+	 */
+	case CUDA_ERROR_INVALID_VALUE:
+		break;
+
+	/* Returned if cuInit() has not been called. This can happen if support
+	 * for CUDA is enabled but the user has not made a CUDA call. This is
+	 * not treated as an error.
+	 */
+	case CUDA_ERROR_NOT_INITIALIZED:
+		break;
+
+	/* Returned if the CUcontext does not support unified virtual
+	 * addressing.
+	 */
+	case CUDA_ERROR_INVALID_CONTEXT:
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"CUcontext does not support unified virtual addressining\n");
+		return -FI_EINVAL;
+
+	default:
+		FI_WARN(&core_prov, FI_LOG_CORE,
+			"Unhandle cuPointerGetAttribute return code: ret=%d\n",
+			cuda_ret);
+		return -FI_EINVAL;
+	}
+#endif /* HAVE_LIBCUDA */
```

Comment:
Let's move the cuda code into the hmem_cuda file.  It seems like we should be able to use the cuda_valid() call for this, if we update that call to accept a NULL hmem_info parameter.

Reply:
I can move this code into hmem_cuda. But, this code is different than the cuda code in cuda_mm_valid(). This function calls cuPointerGetAttribute with CU_POINTER_ATTRIBUTE_MEMORY_TYPE whereas cuda_mm_valid calls cuPointerGetAttribute with CU_POINTER_ATTRIBUTE_BUFFER_ID.